typescript/search-attributes.md

```diff
@@ -0,0 +1,86 @@
+## Search attributes V2
+
+### Why do we need this?
+
+In the current version of the SDK, search attributes are always arrays. In order for the server to support visibility
+on top of standard persistence, we need to redo the search attributes API to enforce single values for all search
+attributes except for keywords. The server will eventually drop support for multiple search attribute values for
+non-keywords.
+
+### Minimal proposal
+
+Make the E2E experience consistent, including query construction for listing workflows and search attribute
+registration.
+
+`workflows.ts`
+
+```ts
+import { searchAttributes } from '@temporalio/common';
+
+export const customerAttribute = searchAttributes.defineAttribute(searchAttributes.Text, 'customer');
+export const dateOfPurchaseAttribute = searchAttributes.defineAttribute(Date, 'dateOfPurchase');
+
+export async function myWorkflow() {
+  // Get the value of a search attribute - returns string | undefined here
+  const customer = customerAttribute.value();
+  // ... random workflow logic ...
+
+  // Set the value of a search attribute, all set and unset calls are buffered and flushed out as a single command at
+  // the end of an activation
+  dateOfPurchaseAttribute.set(new Date());
+  // Explicit unset API
+  dateOfPurchaseAttribute.unset();
+  // ...
+}
+```
+
+`client.ts`
+
+```ts
+import { searchAttributes } from '@temporalio/common';
+import { Client, Q } from '@temporalio/client';
+import { customerAttribute, dateOfPurchaseAttribute, myWorkflow } from './workflows';
+
+const client = new Client(/* options */);
+
+// Only available in OSS for now
+await client.operator.registerSearchAttributes(customerAttribute, dateOfPurchaseAttribute);
+// Other methods will be exposed too
+
+const iterator = client.workflow.list({
+  query: Q.filter(
+    // WorkflowType is a pre-defined search attribute
+    Q.and(searchAttributes.WorkflowType.eq(myWorkflow.name), customerAttribute.eq('customer-id-1234')).orderBy(
+      dateOfPurchaseAttribute,
+      'ASC' // Optionally specify DESC | ASC
+    )
+  ),
+  pageSize: 100, // optional
+});
+
+// Returns something like the TS wrapper returned by Describe - the API returns
+// temporal.api.workflow.v1.WorkflowExecutionInfo
+for await (const { workflowId, ...rest } of iterator) {
+  const handle = client.workflow.getHandle(workflowId);
+  await handle.cancel();
+}
+
+const handle = client.getHandle(someWorkflowId);
+const info = await handle.describe();
+const customer = info.getAttribute(customerAttribute); // includes runtime validation - returns string | undefined
+// -- OR --
+const customer = info.getAttribute<string>(customer); // no runtime validation - returns string | undefined
```

Comment:
I assume you'll do the same for memo here and for the in-workflow part? (though memo can be any type and is not part of your filter work)

Reply:
There's no need to do this for memo, but I do think that the "upsert memo" API in the workflow module will do batching as well.

Reply:
Getting a single memo value by key for a specific type makes sense though, right? `info.getMemo(myStringMemo)` and `info.getMemo<string>("my-string-memo")`

Reply:
Hmm.. right now memos can be any user object so we can't do runtime validation like we do for search attributes.
ATM users can access their memo as: `info.memo['my-string-memo'] as string`.

Adding the getter doesn't provide that much value IMHO but I'll consider adding it for API consistency with search attributes.

Reply:
We added [workflow.memo](https://python.temporal.io/temporalio.workflow.html#memo) and [workflow.memo_value](https://python.temporal.io/temporalio.workflow.html#memo_value) (and the client handle describe response equivalents) after discussion because there is no simple way to deserialize a memo value without manually invoking a converter.

No need to add it now, but I think making memo and search attributes APIs match as much as possible makes sense (though obviously the latter is much more limited on types).

Reply:
If we had type hints in TS I'd do it, but we currently don't support that.